WIP: Setup pack testing

johandahlberg opened the pull request:
Experimentation with setting up auto-tests for this repository. Please note it's not yet ready to be merged.

johandahlberg commented:
This is still work in progress, and at least the following issues need to be resolved before merging:

- [ ] Add a license to this repo
- [ ] Create a script to prepare the proper environment rather than relying in the user running the steps manually first
- [ ] Do no longer download `st2` into `/tmp` when doing manual testing since this mean that it goes away if you restart the machine - once a better location has been determined also fix the hard coded value in `utils/run_tests.sh`
- [ ] Update the README to match changes in the testing process

While this is still a work in progress, it's starting to come together and any feedback is welcome.

johanherman commented on requirements-test.txt:
```diff
@@ -0,0 +1,4 @@
+mock>=1.3.0,<2.0
```
Nice. Didn't know you could specify closed subsets.

johandahlberg commented:
@johanherman Would you like to have a look at this now? I'll probably close this and make a new PR clean up the commit history a bit, but if there's anything I should fix I might as well fix it here.

Of particular interest would be if you could run through the testing instructions in the README and make sure it works as expected for you and that the instructions make sense.

johanherman commented:
Aight. I'll take a look.

johanherman commented on README.md:
```diff
@@ -8,30 +8,32 @@ StackStorm pack to setup automation workflow taking data from the sequencer to d
 If you are using the `arteria-provisioning` this directory should be mounted under the `/opt/stackstorm/packs/arteria-packs` 
```
`arteria-provisioning` is defunct.

johandahlberg replied:
:+1: We should update the repo with a note of this.

johanherman replied:
That's already done since last week.

johanherman commented on .travis.yml:
```diff
@@ -15,8 +15,8 @@ install:
-  - git clone https://github.com/StackStorm/st2.git --depth 1 --single-branch --branch v1.5.1 /tmp/st2 
-  # Super hacky work-around for the problem of the st2.tests.conf having a hard-coded path to the packs in it
-  - mkdir packs && ln -s $PWD packs/arteria-packs
+  - git clone https://github.com/StackStorm/st2.git --depth 1 --single-branch --branch v1.5.1 /tmp/st2
```
Do you have to clone this repo here when you seem to be doing the same in `prepare_test_env.sh`?

johanherman replied:
Never mind. Noticed that I should check the complete changeset and not per commit..

johanherman commented on Vagrantfile:
```diff
@@ -0,0 +1,41 @@
+
+    # NFS-synced directory for pack development
+    # Change "/path/to/directory/on/host" to point to existing directory on your laptop/host and uncomment:
+    config.vm.synced_folder "~/workspace/arteria/arteria-packs", "/arteria-packs", :nfs => true, :mount_options => ['nfsvers=3']
```
I was about to say that you should init this path somehow more conveniently, or update the comment and add some instructions about this path in the README. The init could either be user specified, or Vagrant could find the current working directory and just mount that. 

But - that is actually already happening. Vagrant always mounts the directory where `Vagrantfile` resides, under `/vagrant/` on the guest. Can't you use that instead of mounting under `/arteria-packs`? I notice that you're explicitly specifying NFS v3 -- is that a requirement? I remember you briefly mentioning something about NFS earlier and automatic tests...  


Either this path should be initialized by trying to find e.g. the current working directory (which

johanherman replied:
And if there is a requirement to have it NFS mounted in exactly this way then some further instructions are probably needed, because I'm getting mount problems at the moment when running `vagrant up`: 

```
It appears your machine doesn't support NFS, or there is not an
adapter to enable NFS on this machine for Vagrant. Please verify
that `nfsd` is installed on your machine, and try again. If you're
on Windows, NFS isn't supported. If the problem persists, please
contact Vagrant support.
```

Will look into it a bit more when I've gone through the rest.

johanherman replied:
If you must mount it like this then I'd suggest fetching the current dir with something like `vagrant_root = File.dirname(__FILE__)` and using that instead of `~/workspace/...`

johandahlberg replied:
This was a lazy way to solve that it will use the name of the folder as the name of the package - so this way the it will try to install the package `vagrant`. I'll see if I can resolve it in some nicer way. :smile:

johanherman replied:
It will be OK if you do the symlinks as suggested (and continue to go to `/arteria-packs` as per the instructions).

johanherman commented on Vagrantfile:
```diff
@@ -0,0 +1,41 @@
+      inline: "sudo apt-get install -y curl python-virtualenv vim"
+
+    arteria.vm.provision "shell", 
+      inline: "curl -sSL https://stackstorm.com/packages/install.sh | bash -s -- --user=#{arteriauser} --password=#{arteriapasswd} stable"
```
This will always install the latest stable release (1.6 atm). Do we want this?

johanherman commented on actions/poll_status.yaml:
```diff
@@ -3,7 +3,7 @@ name: poll_status
 description: Polls an URL until the web service says that the long process has finished
 enabled: true
-entry_point: lib/poll_status.py
+entry_point: poll_status.py
```
Is there any specific reasons for why you moved this away from the `lib` subfolder btw?

johandahlberg replied:
It makes it more difficult to make the testing works, since you have to add each folder to the `PYTHONPATH` manually. Moving them into the same folder made this easier.

johanherman replied:
Hm, ok. Seems a bit sad to pollute the folder structure with more messiness than before. :P

johanherman commented on utils/prepare_test_env.sh:
```diff
@@ -0,0 +1,22 @@
+pip install -r requirements-test.txt
+
+# Checkout and install st2 requirements
+git clone https://github.com/StackStorm/st2.git --depth 1 --single-branch --branch v1.5.1 ./st2
```
1. Why are we checking out an other branch here than installed via the St2 bootstrap script? 

2. How can we syncronize this (and the bootstrap install) with the St2 version we actually want to use when we're developing/testing? I'm guessing we would like to have this vs the 1.4 branch atm for example.

johandahlberg replied:
1. Because I botched it. I'll fix that.

2. We need to check it out, because installing and checking out the source serves different purposes. The installation makes it easier to develop things as it means you get a full st2 install available to play around with. The checkout of the source is necessary as it contains testing code which is not part of the packages that gets installed by the install script.

johanherman replied:
Ok. So we need to have both. 

But we still need to have these two versions synced with what we're running at SNPSEQ -- or what we want to run. So at least it must be easy to adjust.

johandahlberg replied:
I've have a solution in the works for this. Stay tuned. :wink:

johanherman commented on Vagrantfile:
```diff
@@ -0,0 +1,41 @@
+
+    # Start shell provisioning.
+    arteria.vm.provision "shell", 
+      inline: "sudo apt-get install -y curl python-virtualenv vim"
```
I recommend doing an `apt-get update` here as well...

johandahlberg replied:
:+1:

johanherman commented on README.md:
```diff
@@ -8,6 +8,40 @@ StackStorm pack to setup automation workflow taking data from the sequencer to d
+./utils/run_tests.sh
+
+# To test registering all pack components run
+./utils/st2-check-register-pack-resources
```
Should be set to e.g. ` ./utils/st2-check-register-pack-resources utils/st2.tests.conf`

johandahlberg replied:
I made it default to that, but perhaps it would be better to be explicit?

johanherman replied:
No, you made it default to `/st2.tests.conf` as far as I can see.

johandahlberg replied:
You're right.

johanherman commented:
Anyways. I ran through the README successfully with some modifications mentioned above. 

I.e. 

- no custom synced folder, 
- `apt-get update` in Vagrantfile, 
- symlink `/arteria-packs` to `/opt/stackstorm/packs/arteria-packs`
- symlink that one to `/vagrant`
- and specify the path to the `st2.tests.conf` file 

There are some other questions and comments above as well though.

johandahlberg commented:
@johanherman I've addressed your comments now. Have a look and once you're happy you can say so, and then I can close and rebase and open a cleaner PR to merge.

johanherman commented on utils/run_tests.sh:
```diff
@@ -0,0 +1,86 @@
+if [  ! -d ${PACK_TESTS_PATH} ]; then
+    echo "Running tests for pack: ${PACK_NAME}"
+    echo "No tests found."
+    exit 0
```
Perhaps you should do an exit != 0 here, as this will pass otherwise as can be seen by the latest travis run.

johanherman commented:
Looks good, although I haven't tested to run it myself with your latest changes. I guess you just have to add a symlink for `/opt/stackstorm/packs/arteria-packs` now to make Travis happy.

johanherman commented:
:+1:

johandahlberg commented:
I'll close and open a new cleaned up one for you to merge.
